syntax = "proto3";

package vocab;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

message VocabRequest {
    string tokenType = 1; // Bearer
    string token = 2;
}

message VocabResponse {
    uint32 id = 1;
    string word = 2;
    string definition = 3;
    google.protobuf.Timestamp createdAt = 4;
    
    bool isLearned = 5;
    bool cards = 6;
    bool wordTranslation = 7;
    bool constructor = 8;
    bool wordAudio = 9;
}

message CreateRequest {
    string word = 1;
    string definition = 2;
    string token = 3;
}

message DeleteRequest {
    string token = 1;
    uint32 word_id = 2;
}

message UpdateRequest {
    uint32 id = 1;
    string definition = 2;
    string token = 3;
}

message UpdateStatusRequest {
    uint32 id = 1;
    bool is_learned = 2;
    string token = 3;
}

message WordRequest {
    uint32 word_id = 1;
}

message ManageTrainingsRequest {
    uint32 id = 1;
    bool res = 2;
    string training = 3;
    string token = 4;
}

message AddWordToStudentRequest {
    string word = 1;
    string definition = 2;
    uint32 user_id = 3;
}

message AddWordToStudentResponse {
    uint32 word_id = 1;
}

message Empty {}

service VocabService {
    rpc AddWordToStudent(AddWordToStudentRequest) returns (AddWordToStudentResponse);
    rpc CreateWord(CreateRequest) returns (google.protobuf.Empty);
    rpc DeleteWord(DeleteRequest) returns (google.protobuf.Empty);
    rpc UpdateWord(UpdateRequest) returns (google.protobuf.Empty);
    rpc UpdateWordStatus(UpdateStatusRequest) returns (google.protobuf.Empty);
    rpc ManageTrainings(ManageTrainingsRequest) returns (google.protobuf.Empty);
    rpc GetWords(VocabRequest) returns (stream VocabResponse);
    rpc FindWord(WordRequest) returns (VocabResponse);
}
